pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/parser/RuntimeASTTransformer.groovy

```diff
@@ -766,30 +815,39 @@ class RuntimeASTTransformer {
      * @return The AST for a constructor call for {@link Matrix}, or the constant null expression if the original
      * cannot be transformed.
      */
+    @Nonnull
     Expression transformMatrix(@CheckForNull ModelASTMatrix original) {
         if (isGroovyAST(original) && !original?.stages?.stages?.isEmpty() && !original?.axes?.axes?.isEmpty()) {
-            ListExpression argList = new ListExpression()
 
             // generate matrix combinations of axes - cartesianProduct
             Set<Map<ModelASTKey, ModelASTValue>> expansion = expandAxes(original.axes.axes)
 
             // remove excluded combinations
             filterExcludes(expansion, original.excludes)
 
-            // for each combination
+            Expression stagesExpression = transformStages(original.stages)
+
+            // With script splitting stagesExpression is a method that creates a new stages expression at runtime
+            // If disabled, we still want to make this a script closure call so we can reuse it instead of generating code for every cell
+            if (!SCRIPT_SPLITTING_TRANSFORMATION) {
+                stagesExpression = wrapper.asWrappedScriptContextVariable(stagesExpression, true)
+            }
```

Comment:
@jglick  Per your comment, this forces the use of a single expression (method or a closure), which is then called for each cell instead of generating code for each cell.